common/workunit/workunit.cpp

```diff
@@ -6805,6 +6860,80 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
     throwUnexpected();
 }
 
+//Verifies the given signed workunit distributed access token was created
+//and signed by the given wuid and user, and that the workunit is still active
+//Returns:
+//   0 : Success, token is valid and workunit is active
+//   1 : Signature does not verify (wuid/username don't match, or signature doesnt verify)
+//   2 : Workunit not active
+//   Throws if unable to open workunit
+CriticalSection verifyWorkunitDATokenCrit;
+int verifyWorkunitDAToken(const char * wuid, const char * user, const char * distributedAccessToken)
+{
+    CriticalBlock block(verifyWorkunitDATokenCrit);
+
+    //Validate user
+    const char * finger = strstr(distributedAccessToken, "u=");
+    StringBuffer sb;
+    if (finger)
+    {
+        finger += 2;
+        while (*finger && *finger != ',' && *finger != ']')
+            sb.append(1, (const char *)finger++);
+    }
+    if (0 != strcmp(user, sb.str()))
+    {
+        ERRLOG("WUDistributed Access Token user does not match given user");
+        return 1;
+    }
+
+    //Validate wuid
+    finger = strstr(distributedAccessToken, "w=");
+    sb.clear();
+    if (finger)
+    {
+        finger += 2;
+        while (*finger && *finger != ',' && *finger != ']')
+            sb.append(1, (const char *)finger++);
+    }
+    if (0 != strcmp(wuid, sb.str()))
+    {
+        ERRLOG("WUDistributed Access Token wuid does not match given wuid");
+        return 1;
+    }
+
+    //Validate signature
+    IDigitalSignatureManager * pDSM = queryDigitalSignatureManagerInstanceFromEnv();
+    if (pDSM && pDSM->isDigiVerifierConfigured())
+    {
+        StringBuffer token;//receives copy of everything up until signature
+        for (finger = distributedAccessToken; *finger && *finger != ';'; finger++)
+            token.append(1, finger);
+        token.append(1, finger);//append ;
+
+        if (finger && *finger)
+        {
+            StringBuffer sig(++finger);
+            if (!pDSM->digiVerify(token, sig))
+            {
+                ERRLOG("WUDistributed Access Token signature does not verify");
+                return 1;
+            }
+        }
+    }
+
+    //Verify workunit still active
+    Owned<IWorkUnitFactory> factory = getWorkUnitFactory();
+    Owned<IConstWorkUnit> cw = factory->openWorkUnit(wuid);
+    if(!cw)
+    {
+        throw MakeStringException(WUERR_WorkunitAccessDenied,"Cannot open workunit %s",wuid);
+    }
+
+//TODO Should I allow other states (compiling, submitted, paused etc)???
```

Comment:
I think for now only WUStateRunning and if it's ever used WUStateDebugRunning.  That's the only time the workunit should be calling out to services.  If the compilation servers need to do anything on behalf of the workunit we should consider that carefully and therefore if it ends up being token based, we should break it until we know we need to allow it.